File: app/src/main/java/com/ladsoft/bakingapp/translator/RecipeTranslator.java
package com.ladsoft.bakingapp.translator;

import com.ladsoft.bakingapp.data.network.entity.IngredientPayload;
import com.ladsoft.bakingapp.data.network.entity.RecipePayload;
import com.ladsoft.bakingapp.data.network.entity.StepPayload;
import com.ladsoft.bakingapp.entity.Ingredient;
import com.ladsoft.bakingapp.entity.Recipe;
import com.ladsoft.bakingapp.entity.Step;

import java.util.ArrayList;
import java.util.List;

public class RecipeTranslator implements Translator<RecipePayload, Recipe> {
    @Override
    public Recipe translate(RecipePayload source) {
        IngredientTranslator ingredientTranslator = new IngredientTranslator(source.getId());
        StepTranslator stepTranslator = new StepTranslator(source.getId());

        List<Ingredient> ingredients = new ArrayList<>();
        List<Step> steps = new ArrayList<>();

        List<IngredientPayload> ingredientsPayload = source.getIngredients();
        if (ingredientsPayload != null) {
            long incrementalId = 0;
            for (IngredientPayload ingredientPayload : ingredientsPayload) {
                ingredientPayload.setId(++incrementalId);
                ingredients.add(ingredientTranslator.translate(ingredientPayload));
            }
        }

        List<StepPayload> stepsPayload = source.getSteps();
        if (stepsPayload != null) {
            for (StepPayload stepPayload : stepsPayload) {
                steps.add(stepTranslator.translate(stepPayload));
            }
        }

        return new Recipe(source.getId(),
                        source.getName() != null ? source.getName() : "",
                        source.getServings(),
                        source.getImageUrl() != null ? source.getImageUrl() : "",
                        ingredients,
                        steps);
    }
}

File: app/src/main/java/com/ladsoft/bakingapp/translator/RecipeTranslator.kt
package com.ladsoft.bakingapp.translator

import com.ladsoft.bakingapp.data.network.entity.RecipePayload
import com.ladsoft.bakingapp.entity.Recipe

class RecipeTranslator : Translator<RecipePayload, Recipe> {
    override fun translate(source: RecipePayload): Recipe {
        val ingredientTranslator = IngredientTranslator(source.id)
        val stepTranslator = StepTranslator(source.id)

        val ingredients = source.ingredients
                ?.sortedWith(compareBy{it.description?.toUpperCase()})
                ?.mapIndexed { index, it ->
                    it.id = index.toLong()
                    ingredientTranslator.translate(it)
                }

        val steps = source.steps?.map { stepTranslator.translate(it) }

        return Recipe(source.id, source.name ?: "", source.servings, source.imageUrl ?: "",
                    ingredients ?: emptyList(), steps ?: emptyList())
    }
}
